### PR TITLE
feat(tianmu): refactor tianm and fix warnings(#11)

### PR DESCRIPTION
**Pull request by RingsC**
## Summary about this PR

[summary]
1: refactor tianmu code.
2: fix the warnings.

Issue Number: close #11

## Tests Check List

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [ ] New Feature
- [ ] Bug Fix
- [ ] Improvement
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

File: storage/tianmu/core/multi_index.h [hunk 1]
```diff
@@ -157,33 +173,35 @@ class MultiIndex {
...
-                      // updating iterator exists
+  std::vector<bool> can_be_distinct_;  // true if the dimension contain only one copy of
+                                       // original rows, false if we cannot guarantee this
+  std::vector<bool> used_in_output_;   // true if given dimension is used for
+                                       // generation of output columns
+
+  std::vector<DimensionGroup *> dim_groups_;  // all active dimension groups
+  DimensionGroup **dimension_group_;          // pointers to elements of dim_groups, for
+                                              // faster dimension identification
+  int *group_num_for_dimension_;              // an element number of dim_groups, for faster
```
Comment (lujiashun) [inline storage/tianmu/core/multi_index.h]: @hustjieke  prefers to initialize the class member variables( non-class) once they are declared.  Should we uniform the format style?

File: storage/tianmu/core/multi_index.h [hunk 2]
```diff
@@ -157,33 +173,35 @@ class MultiIndex {
...
-  DimensionGroup **group_for_dim;            // pointers to elements of dim_groups, for
-                                             // faster dimension identification
-  int *group_num_for_dim;                    // an element number of dim_groups, for faster
-                                             // dimension identification
 
-  // Some technical functions
-  void MultiplyNoTuples(uint64_t factor);  // the same as "no_tuples*=factor", but set
-                                           // no_tuples_too_big whenever needed
+  uint32_t num_of_dimensions_{0};
+  int64_t *dim_size_{0};       // the size of a dimension
```
Comment (lujiashun) [inline storage/tianmu/core/multi_index.h]: Better to use nullptr for a pointer

**Review by lujiashun — APPROVED** [review]
LGTM

**Review by DandreChen — APPROVED** [review]
LGTM.

**Review by hustjieke — APPROVED** [review]
LGTM

**Pull request merged**